dsx-fw/DSX_Library/Command/Src/validate.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * validate.c
+ *
+ *  Created on: May 26, 2021
+ *      Author: jpaca
+ */
+
+#include "validate.h"
+
+
+// Validate Digital Write Command
+bool validate_digital_write(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc != DO1 && dsx_data->loc != DO2) check_data = false;
+	if(dsx_data->sign != POSITIVE) check_data = false;
+	if(dsx_data->val != 0 && dsx_data->val != 1) check_data = false;
+//	if(dsx_data->ret != CMD_COMPLETE_PING) check_data = false;
+	return check_data;
+}
+
+// Validate Digital Read Command
+bool validate_digital_read(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc < DI1 || dsx_data->loc > DI8) check_data = false;
+	return check_data;
+}
+
+// Validate Get Pin Mode Command
+bool validate_get_pin_mode(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc < DI1 || dsx_data->loc > AO2) check_data = false;
+	return check_data;
+}
+
+// Validate Analog Read Command
+bool validate_analog_read(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc < AI1 || dsx_data->loc > AI4) check_data = false;
+	return check_data;
+}
+
+// Validate PWM Write Command
+bool validate_pwm_write(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc != PWM1 && dsx_data->loc != PWM2) check_data = false;
+	if(dsx_data->sign != POSITIVE) check_data = false;
+	if(dsx_data->val < DUTYCYCLE_MIN || dsx_data->val > DUTYCYCLE_MAX) check_data = false;
+//	if(dsx_data->ret != CMD_COMPLETE_PING) check_data = false;
+	return check_data;
+}
+
+// Validate Set PWM Frequency Command
+bool validate_set_PWM_freq(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc != PWM1 && dsx_data->loc != PWM2) check_data = false;
+	if(dsx_data->sign != 0 && dsx_data->sign != 1) check_data = false;
+
+	// Sign = 1: 26-9999 Hz, Sign = 2: dsx_data->val = 1000-2000 + dsx_data->ret = 0-9
+	if(dsx_data->sign == 0){
+		if(dsx_data->val < PWM_MIN || dsx_data->val > 9999) check_data = false;
+	}
+	if(dsx_data->sign == 1){
+		if(dsx_data->val < 1000 || dsx_data->val > 2000) check_data = false;
+		if(dsx_data->ret < 0 || dsx_data->ret > 9) check_data = false;
+	}
+	return check_data;
+}
+
+// Validate Servo Write Command
+bool validate_servo_write(volatile DSX_data_t *dsx_data){
+	bool check_data = true;
+	if(dsx_data->loc != PWM1 && dsx_data->loc != PWM2) check_data = false;
+	if(dsx_data->sign != POSITIVE) check_data = false;
+
+	// Define max and min degree macros for servo in the servo library
+	if(dsx_data->val < MIN_DEGREES || dsx_data->val > MAX_DEGREES) check_data = false;
+//	if(dsx_data->ret != CMD_COMPLETE_PING) check_data = false;
+	return check_data;
+}
+
+// Validate Read Encoder Speed and Direction Command
+bool validate_read_encoder(volatile DSX_data_t *dsx_data){};
```

Comment:
You forgot to add logic for validating a command to read the encoder

Reply:
I added validation for the sign but do we need to check if the speed does not exceed some maximum value?

Reply:
I looked at the lookup table for encoder read. It looks like every parameter except ID is don't care so you don't need to check for anything. 

You don't need to check the speed because speed is not sent in the command to read the encoder. Simply returning true is okay in validate_read_encoder() based on the lookup table for encoder read.

Reply:
This is why the validation was left empty cause there was nothing to check for. There aren't options for encoder ports as the two pins are only dedicated to the encoder so I don't think we need to checks for the locs. Also there are no board defines for the encoder, at least in my branch.

Reply:
@Harjaapb validate_read_encoder needs to return true. or we'll never be able to read the encoder, that's what I'm saying.

Reply:
They way you have it, validate_read_encoder will always return false.